### ISSUE TITLE
fp.rem constant-folding float/double performance

### ISSUE THREAD

**Issue by LeventErkok** [Z3Prover/z3#96]
I have 2 identical benchmarks for computing (i.e., constant-folding) `fp.rem NaN NaN`; one for single-precision and the other for double-precision.

Turns out z3 can answer the query on single-precision in about 9 seconds on my (somewhat dated) computer. Rather slow for constant folding; but perhaps acceptable given all the complexity associated with FPA.

But z3 simply does not complete the`Double` instance at all: I had to kill it about 5 mins into the run-time.

Both benchmarks are included below; first for `Float` and then for `Double`. Perhaps there's a performance issue around constant-folding with floats/doubles that really throw Z3 off in case of double-precision numbers?

`Float` version; z3 answers it in about 9 seconds on my computer:

```
; Automatically generated by SBV. Do not edit.
(set-option :produce-models true)
(set-logic QF_FP)
; --- uninterpreted sorts ---
; --- literal constants ---
(define-fun s_2 () Bool false)
(define-fun s_1 () Bool true)
; --- skolem constants ---
(declare-fun s0 () (_ FloatingPoint  8 24))
(declare-fun s1 () (_ FloatingPoint  8 24))
(declare-fun s2 () (_ FloatingPoint  8 24))
; --- constant tables ---
; --- skolemized tables ---
; --- arrays ---
; --- uninterpreted constants ---
; --- user given axioms ---
; --- formula ---
(assert ; no quantifiers
   (let ((s3 (fp.isNaN s0)))
   (let ((s4 (fp.isNaN s1)))
   (let ((s5 (fp.rem s0 s1)))
   (let ((s6 (= s2 s5)))
   (let ((s7 (and s4 s6)))
   (let ((s8 (and s3 s7)))
   s8)))))))
(check-sat)
(get-model)
```

`Double` version; I killed the process after about 5 minutes of running:

```
; Automatically generated by SBV. Do not edit.
(set-option :produce-models true)
(set-logic QF_FP)
; --- uninterpreted sorts ---
; --- literal constants ---
(define-fun s_2 () Bool false)
(define-fun s_1 () Bool true)
; --- skolem constants ---
(declare-fun s0 () (_ FloatingPoint 11 53))
(declare-fun s1 () (_ FloatingPoint 11 53))
(declare-fun s2 () (_ FloatingPoint 11 53))
; --- constant tables ---
; --- skolemized tables ---
; --- arrays ---
; --- uninterpreted constants ---
; --- user given axioms ---
; --- formula ---
(assert ; no quantifiers
   (let ((s3 (fp.isNaN s0)))
   (let ((s4 (fp.isNaN s1)))
   (let ((s5 (fp.rem s0 s1)))
   (let ((s6 (= s2 s5)))
   (let ((s7 (and s4 s6)))
   (let ((s8 (and s3 s7)))
   s8)))))))
(check-sat)
(get-model)
```

Note that the benchmarks are identical except the first uses `(_ FloatingPoint 8 24)` while the second uses `(_ FloatingPoint 11 53)`.

**Comment by wintersteiger** [comment]
This is known and expected. Currently these operations are bit-blasted (eagerly) and they produce huge SAT problems, in the case of double precision fp.rem, my desktop runs out of memory during bit-blasting, before it does any solving. There is a time/memory trade-off to be considered, but so far I haven't seen any meaningful practical application of this operation, so I'm reluctant to invest effort on turning memory-outs into time-outs which ultimately doesn't help anybody.

(Aside: This is also the implicit reason why many CPUs (including Intel x86, x64), don't offer a full fp.rem instruction; the circuit would be too large. They only offer two versions of partial remainder operations, FPREM and FPREM1, which can be used to get precise results through iteration, i.e., they are also trading time for memory.)

**Comment by wintersteiger** [comment]
Closing this issue as "won't be fixed without further motivation".

Note that `(fp.isNaN s0)` does not trivially mean that s0 is a constant/value. So, the constant/value propagation will not pick it up. If we replace that with `(= s0 (_ NaN 11 53))` it sees a proper value and it will propagate it, so that this particular benchmark is solved by the simplifier. I've added automatic injection of those auxiliary lemmas to the fpa2bv converter, but this obviously works only for isNaN, not for any of the other fp.is* predicates or other values.
